File: lib/moysklad/transform.ts
/**
 * Преобразование сырых данных МойСклад в формат для аналитики.
 */

import type { InventoryInput } from '../analytics/inventory';
import type { AbcInput } from '../analytics/abc';
import type { XyzInput } from '../analytics/xyz';
import type { RfmTransaction } from '../analytics/rfm';
import type { MsAssortmentItem, MsAttribute, MsDemand } from './types';

/**
 * Из ассортимента + истории отгрузок собирает входы для inventory-анализа.
 * Средние дневные продажи считаются по фактическим отгрузкам за период.
 *
 * Если в МойСклад на товаре задано дополнительное поле «норматив запаса»,
 * передайте его имя или ID в `normDaysAttribute` — значение возьмётся оттуда.
 * Иначе используется `defaultNormDays`.
 */
/** Курс валюты к базовой: множитель = rate / multiplicity (база = 1). */
export type CurrencyRate = { symbol: string; rate: number; multiplicity: number };

export function assortmentToInventory(
  items: MsAssortmentItem[],
  demands: MsDemand[],
  opts: {
    periodDays: number;
    defaultNormDays?: number;
    priceTypeName?: string;
    normDaysAttribute?: string; // имя или ID кастомного атрибута
    /** Карта currencyId → { symbol, rate, multiplicity } для конвертации в базовую валюту. */
    currencyByHref?: Map<string, CurrencyRate>;
  } = { periodDays: 30 },
): InventoryInput[] {
  const fallbackNorm = opts.defaultNormDays ?? 10;
  const salesByProduct = new Map<string, number>();
  for (const d of demands) {
    for (const p of d.positions?.rows ?? []) {
      const id = extractAssortmentId(p.assortment?.meta?.href);
      if (!id) continue;
      salesByProduct.set(id, (salesByProduct.get(id) ?? 0) + p.quantity);
    }
  }

  const lookupRate = (href: string | undefined): CurrencyRate | undefined => {
    if (!href || !opts.currencyByHref) return undefined;
    const id = extractAssortmentId(href);
    if (!id) return undefined;
    return opts.currencyByHref.get(id);
  };
  // Коэффициент перевода в базовую валюту. База имеет rate=1, mult=1 → 1.
  const toBase = (cur: CurrencyRate | undefined): number =>
    cur && cur.rate > 0 && cur.multiplicity > 0 ? cur.rate / cur.multiplicity : 1;

  return items.map((it) => {
    const buyRate = lookupRate(it.buyPrice?.currency?.href);
    const salePicked = pickSalePrice(it, opts.priceTypeName);
    const saleRate = lookupRate(salePicked.currencyHref);

    // Исходные цены в валюте карточки
    const costOriginal = (it.buyPrice?.value ?? 0) / 100;
    const saleOriginal = salePicked.value / 100;

    // Конвертируем в базовую валюту аккаунта — только так маржа/наценка корректны
    const buyMul = toBase(buyRate);
    const saleMul = toBase(saleRate);
    const cost = costOriginal * buyMul;
    const sale = saleOriginal * saleMul;
    const converted = buyMul !== 1 || saleMul !== 1;

    const sold = salesByProduct.get(it.id) ?? 0;
    const avgDaily = opts.periodDays > 0 ? sold / opts.periodDays : 0;
    const normDays = extractNormDays(it.attributes, opts.normDaysAttribute, fallbackNorm);
    return {
      id: it.id,
      name: it.name,
      stock: it.stock ?? 0,
      costPrice: cost,
      salePrice: sale,
      avgDailySales: avgDaily,
      normDays,
      buyCurrency: buyRate?.symbol,
      saleCurrency: saleRate?.symbol,
      costPriceOriginal: costOriginal,
      salePriceOriginal: saleOriginal,
      converted,
    };
  });
}

/**
 * Достаёт значение норматива из массива атрибутов. Атрибут можно указать
 * либо по UUID, либо по имени (точное совпадение, потом без регистра).
 */
export function extractNormDays(
  attributes: MsAttribute[] | undefined,
  attrKey: string | undefined,
  fallback: number,
): number {
  if (!attrKey || !attributes?.length) return fallback;
  const match =
    attributes.find((a) => a.id === attrKey) ??
    attributes.find((a) => a.name === attrKey) ??
    attributes.find((a) => a.name.toLowerCase() === attrKey.toLowerCase());
  if (!match) return fallback;
  const v = match.value;
  if (typeof v === 'number' && Number.isFinite(v) && v > 0) return v;
  if (typeof v === 'string') {
    const n = parseFloat(v.replace(',', '.'));
    if (Number.isFinite(n) && n > 0) return n;
  }
  return fallback;
}

function pickSalePrice(
  item: MsAssortmentItem,
  priceTypeName?: string,
): { value: number; currencyHref?: string } {
  const prices = item.salePrices ?? [];
  if (priceTypeName) {
    const match = prices.find((p) => p.priceType?.name === priceTypeName);
    if (match) return { value: match.value, currencyHref: match.currency?.href };
  }
  const first = prices[0];
  return { value: first?.value ?? 0, currencyHref: first?.currency?.href };
}

function extractAssortmentId(href: string | undefined): string | null {
  if (!href) return null;
  const m = href.match(/\/([0-9a-f-]{36})(?:\/|$|\?)/i);
  return m ? m[1] : null;
}

/** Группирует отгрузки по позициям и возвращает выручку по каждому SKU */
export function demandsToAbc(demands: MsDemand[]): AbcInput[] {
  const map = new Map<string, { name: string; value: number }>();
  for (const d of demands) {
    for (const p of d.positions?.rows ?? []) {
      const id = extractAssortmentId(p.assortment?.meta?.href);
      if (!id) continue;
      const value = (p.price * p.quantity) / 100;
      const prev = map.get(id);
      if (prev) {
        prev.value += value;
      } else {
        map.set(id, { name: p.assortment?.name ?? id, value });
      }
    }
  }
  return [...map.entries()].map(([id, v]) => ({ id, name: v.name, value: v.value }));
}

/**
 * Группирует отгрузки по SKU и недельным/месячным бакетам.
 * Возвращает массивы продаж по периодам для XYZ.
 */
export function demandsToXyz(
  demands: MsDemand[],
  opts: { bucketDays: number; periodsCount: number; until: Date },
): XyzInput[] {
  const { bucketDays, periodsCount, until } = opts;
  const map = new Map<string, { name: string; periods: number[] }>();

  for (const d of demands) {
    const moment = new Date(d.moment);
    if (!Number.isFinite(moment.getTime())) continue;
    const ageDays = Math.max(
      0,
      Math.floor((until.getTime() - moment.getTime()) / (1000 * 60 * 60 * 24)),
    );
    const bucket = Math.floor(ageDays / bucketDays);
    if (bucket >= periodsCount) continue;
    const idx = periodsCount - 1 - bucket; // более старые слева

    for (const p of d.positions?.rows ?? []) {
      const id = extractAssortmentId(p.assortment?.meta?.href);
      if (!id) continue;
      let row = map.get(id);
      if (!row) {
        row = { name: p.assortment?.name ?? id, periods: new Array(periodsCount).fill(0) };
        map.set(id, row);
      }
      row.periods[idx] += p.quantity;
    }
  }

  return [...map.entries()].map(([id, v]) => ({ id, name: v.name, periods: v.periods }));
}

/** Каждая отгрузка → одна RFM-транзакция (агент = клиент) */
export function demandsToRfm(demands: MsDemand[]): RfmTransaction[] {
  return demands
    .filter((d) => d.agent?.meta?.href)
    .map((d) => ({
      customerId: extractAssortmentId(d.agent!.meta.href) ?? d.agent!.meta.href,
      customerName: d.agent?.name,
      date: d.moment,
      amount: d.sum / 100,
    }));
}

File: backend/app/moysklad/transform.py
"""
Преобразование сырых данных МойСклад в формат для аналитики.
Порт widget/lib/moysklad/transform.ts. Реализовано:
  - assortment_to_inventory
  - demands_to_abc
  - demands_to_xyz
  - demands_to_rfm
Customer segments (статусы контрагентов из МойСклад) — Phase 3.
"""
from __future__ import annotations

import re
from datetime import datetime
from typing import Any, Optional

from ..analytics.abc import AbcInput
from ..analytics.inventory import InventoryInput
from ..analytics.rfm import RfmTransaction
from ..analytics.xyz import XyzInput

_UUID_RE = re.compile(r"([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})", re.I)


def extract_uuid(href: Optional[str]) -> Optional[str]:
    if not href:
        return None
    m = _UUID_RE.search(href)
    return m.group(1) if m else None


def assortment_to_inventory(
    items: list[dict],
    demands: list[dict],
    period_days: int,
    default_norm_days: float = 10,
) -> list[InventoryInput]:
    """
    Упрощённая версия (без курсов валют и FIFO — это Phase 3).
    avg_daily_sales считается из суммарного количества в отгрузках за период.
    """
    sales_by_product: dict[str, float] = {}
    for d in demands:
        positions = (d.get("positions") or {}).get("rows", []) or []
        for p in positions:
            href = ((p.get("assortment") or {}).get("meta") or {}).get("href")
            pid = extract_uuid(href)
            if not pid:
                continue
            sales_by_product[pid] = sales_by_product.get(pid, 0) + p.get("quantity", 0)

    result: list[InventoryInput] = []
    for it in items:
        item_id = it.get("id")
        if not item_id:
            continue
        cost_kopecks = (it.get("buyPrice") or {}).get("value", 0) or 0
        sale_kopecks = 0
        for sp in it.get("salePrices") or []:
            sale_kopecks = sp.get("value", 0) or 0
            break  # берём первую цену продажи; Phase 3 — выбор по priceType
        sold = sales_by_product.get(item_id, 0)
        avg_daily = (sold / period_days) if period_days > 0 else 0

        result.append(InventoryInput(
            id=item_id,
            name=it.get("name", item_id),
            stock=it.get("stock", 0) or 0,
            cost_price=cost_kopecks / 100,   # МойСклад хранит цены в копейках
            sale_price=sale_kopecks / 100,
            avg_daily_sales=avg_daily,
            norm_days=default_norm_days,
        ))
    return result


def demands_to_abc(demands: list[dict]) -> list[AbcInput]:
    """
    Сумма по позициям (после скидки) в копейках → делим на 100.
    Точно так же как в TS-версии: price * quantity * (1 - discount/100) / 100.
    """
    agg: dict[str, dict[str, Any]] = {}
    for d in demands:
        positions = (d.get("positions") or {}).get("rows", []) or []
        for p in positions:
            href = ((p.get("assortment") or {}).get("meta") or {}).get("href")
            pid = extract_uuid(href)
            if not pid:
                continue
            discount = max(0.0, min(100.0, float(p.get("discount") or 0))) / 100.0
            value = (p.get("price", 0) * p.get("quantity", 0) * (1 - discount)) / 100.0
            name = (p.get("assortment") or {}).get("name") or pid
            row = agg.get(pid)
            if row:
                row["value"] += value
            else:
                agg[pid] = {"name": name, "value": value}

    return [AbcInput(id=pid, name=row["name"], value=row["value"]) for pid, row in agg.items()]


def demands_to_xyz(
    demands: list[dict],
    until: datetime,
    bucket_days: int = 7,
    periods_count: int = 8,
) -> list[XyzInput]:
    """
    Группирует отгрузки по SKU и недельным бакетам. Старые слева, новые
    справа — как в TS-версии.
    """
    agg: dict[str, dict[str, Any]] = {}
    until_ts = until.timestamp()
    for d in demands:
        moment_str = d.get("moment")
        if not moment_str:
            continue
        try:
            # МойСклад отдаёт moment в формате "2024-05-13 10:30:00.000"
            # либо ISO. Парсим оба варианта.
            moment_dt = _parse_ms_moment(moment_str)
        except Exception:
            continue
        age_days = max(0, int((until_ts - moment_dt.timestamp()) // 86400))
        bucket = age_days // bucket_days
        if bucket >= periods_count:
            continue
        idx = periods_count - 1 - bucket  # старые слева

        positions = (d.get("positions") or {}).get("rows", []) or []
        for p in positions:
            href = ((p.get("assortment") or {}).get("meta") or {}).get("href")
            pid = extract_uuid(href)
            if not pid:
                continue
            row = agg.get(pid)
            if not row:
                row = {"name": (p.get("assortment") or {}).get("name") or pid,
                       "periods": [0.0] * periods_count}
                agg[pid] = row
            row["periods"][idx] += p.get("quantity", 0)

    return [XyzInput(id=pid, name=row["name"], periods=row["periods"]) for pid, row in agg.items()]


def _parse_ms_moment(s: str) -> datetime:
    """
    МойСклад moment: "2024-05-13 10:30:00.000" (с пробелом, без таймзоны
    — Moscow). На крайний случай поддерживаем ISO. Тз игнорируем —
    моменты сравниваем с naive `until` через timestamp.
    """
    try:
        return datetime.fromisoformat(s.replace(" ", "T"))
    except ValueError:
        # Если пришёл без миллисекунд
        return datetime.strptime(s, "%Y-%m-%d %H:%M:%S")


def demands_to_rfm(demands: list[dict]) -> list[RfmTransaction]:
    """
    Каждая отгрузка → одна RFM-транзакция (agent = клиент). Порт
    demandsToRfm из TS. Используем d.sum (в копейках) делим на 100 →
    базовая валюта аккаунта. Демонстрации без agent отбрасываем —
    встречаются на retaildemand с анонимной розничной продажей.
    """
    out: list[RfmTransaction] = []
    for d in demands:
        agent = d.get("agent") or {}
        href = (agent.get("meta") or {}).get("href")
        if not href:
            continue
        customer_id = extract_uuid(href) or href
        moment_str = d.get("moment")
        if not moment_str:
            continue
        try:
            date = _parse_ms_moment(moment_str)
        except Exception:
            continue
        out.append(RfmTransaction(
            customer_id=customer_id,
            customer_name=agent.get("name"),
            date=date,
            amount=(d.get("sum", 0) or 0) / 100,
        ))
    return out
